Track ownership for scale subresource of deployments

nodo opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Track ownership for scale subresource of deployments. This is achieved using the main resource field manager.

**Which issue(s) this PR fixes**:

Part of #82046

**Special notes for your reviewer**:

Passing the main resource field manager using context is not great, however it could be an acceptable compromise given we might refactor this logic in https://github.com/kubernetes/kubernetes/issues/84530.

This PR uses some pieces of code and suggestions from https://github.com/kubernetes/kubernetes/pull/83294 and https://github.com/kubernetes/kubernetes/pull/83444.

**Does this PR introduce a user-facing change?**:

```release-note
Track ownership when changing replicas of a deployment via the scale subresource
```

nodo committed:
Track ownership for scale subresource of deployments

nodo commented:
/cc @apelisse @julianvmodesto

fedebongio commented:
/assign @jennybuckley 
/triage accepted

julianvmodesto commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2811,3 +2811,187 @@ spec:
+	assertCompleteOwnership(t, (*deployment).GetManagedFields(), "replace_test")
+}
+
+func deploymentWithReplicas(replicas int) []byte {
```
nit: maybe inline this in the test? It looks like we use the same object several times.

nodo replied:
I agree, it would be cleaner 👍

julianvmodesto commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2811,3 +2811,187 @@ spec:
+	}`, replicas))
+}
+
+func assertCompleteOwnership(t *testing.T, managedFields []metav1.ManagedFieldsEntry, fieldManager string) {
```
Hm I wonder if, instead of the helper, it's better to get the managed fields at an index `deployment.GetManagedFields()[i]`, assert the manager name, and assert that it contains the `.spec.replicas` field.

But for the helper:
- this probably could be named assertReplicasOwnership instead
- I think we want `if !managerSeen { t.Fatalf(...) }` at the end if `.spec.replicas` is not in managed fields at all for some reason

nodo replied:
IIRC I think the reason we have this helper is to ensure that a new manager is owning the replicas field, but also no other manager is, right? I am not sure using `deployment.GetManagedFields()[i]` we could check both conditions.

As for the suggestions, totally! thanks for spotting, I will make the changes

julianvmodesto commented:
Nice -- could you add/update a unit test for this in addition to the integration test? I think in storage_test.go.

julianvmodesto commented:
/ok-to-test

nodo commented:
> Nice -- could you add/update a unit test for this in addition to the integration test? I think in storage_test.go.

@julianvmodesto I think this is much harder that it seems. Here is what I did:

1. Just before [this line](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/deployment/storage/storage_test.go#L267) I have needed to add a field manager to `ctx`
2. I wanted to re-use the field manager used the other unit tests. However, most of the helper functions are in a `_test.go` file, which prevent them from being imported. In other words I had to add [these components](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go#L47-L176) and [this one](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/skipnonapplied_test.go#L31-L43) to a separate `testing.go` file in the same package.
3. In addition, the open-api spec are loaded using a relative path ([see here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go#L47-L51) ) and I had hack my way around it.
4. After, all these changes the result is this commit: https://github.com/nodo/kubernetes/commit/ce0975981f064c161f6cfae3296991388695af12

...which doesn't work (yet!) :-)

Here is the error I am getting:
```
=== RUN   TestScaleUpdate
I1108 14:21:39.022246   39678 client.go:360] parsed scheme: "endpoint"
I1108 14:21:39.022328   39678 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:28500059854110329640  <nil> 0 <nil>}]
I1108 14:21:39.024144   39678 client.go:360] parsed scheme: "endpoint"
I1108 14:21:39.024170   39678 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:28500059854110329640  <nil> 0 <nil>}]
I1108 14:21:39.024319   39678 once.go:66] CPU time info is unavailable on non-linux or appengine environment.
I1108 14:21:39.025119   39678 client.go:360] parsed scheme: "endpoint"
I1108 14:21:39.025146   39678 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:28500059854110329640  <nil> 0 <nil>}]
    storage_test.go:297: error updating scale {{ } {foo-deployment  foo-namespace    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {12} {0 }}: failed to convert new object (/, Kind=) to smd typed: no corresponding type for /, Kind=
--- FAIL: TestScaleUpdate (0.62s)
FAIL
FAIL    k8s.io/kubernetes/pkg/registry/apps/deployment/storage  3.948s
FAIL
```
After some digging, I think I have narrowed down the problem to the [fakeObjectConvertor](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go#L68-L70). However, after some more investigation I wasn't convinced this was worth the effort. 

What do you think?

nodo commented:
Not sure why tests are suddenly failing. I have also tried to rebase but nothing, trying to re-test.

/retest

BenTheElder commented:
/test all

spiffxp commented:
I tried manually running the same commands as listed in https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/95921/pull-kubernetes-typecheck/1326247824702050304/build-log.txt and got the same result

```
# ...
$ git fetch https://github.com/kubernetes/kubernetes.git pull/95921/head
$ git merge --no-ff 92383b25ba1455d2ddc866a57b1b7e249c9b04f8
merge: 92383b25ba1455d2ddc866a57b1b7e249c9b04f8 - not something we can merge
```

Debugging...
```
$ git log -n1 92383b25ba1455d2ddc866a57b1b7e249c9b04f8
fatal: bad object 92383b25ba1455d2ddc866a57b1b7e249c9b04f8
$ cat .git/FETCH_HEAD
1a439c03e242a0735f53bf54e0ac2c999844d836		'refs/pull/95921/head' of https://github.com/kubernetes/kubernetes
$ hub api /repos/kubernetes/kubernetes/pulls/95921 | jq -r .head.sha
92383b25ba1455d2ddc866a57b1b7e249c9b04f8
```

Looks like GitHub's API is lying to us.  I'm not sure if this is an unlucky pull request or if there are known issues with GitHub at present.

I'm not knowledgeable enough about the different use cases pod-utils serve to understand if they should be re-implemented to ignore GitHub's API

neolit123 commented:
> Looks like GitHub's API is lying to us

does seem like so.
i cannot find results in other github projects having this problem from today.

spiffxp commented:
possible github issue, please link if you run into this on other PR's https://github.com/kubernetes/test-infra/issues/19910

julianvmodesto commented:
Oh yes I understand, sorry about that and thanks for trying this!

Yeah, the change looks good and the integration test is good coverage for the wiring here. I don't think it's worth it to block on wiring the unit tests.

/lgtm

julianvmodesto commented:
/assign @apelisse

/test all

nodo commented:
Sorry for the noise, I had to push an empty commit and rebase to fix the problem with Github API. Thanks for your help @spiffxp @neolit123 @BenTheElder !

@julianvmodesto - thanks! please could you have another look? The `lgtm` got removed because of the above.

julianvmodesto commented:
/lgtm

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:
```diff
@@ -604,6 +604,26 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
+				return nil, nil, err
+			}
+
+			reqScope.ParentFieldManager, err = fieldmanager.NewDefaultFieldManager(
```
I typically try to avoid having variables that only make sense in a specific context and not others. In that case, the ParentFieldManager member is always in scopes but is only used for specific sub-resource. In that case (scale sub-resource), do we actually use both fieldmanagers? Is it possible that the fieldmanager above is misconfigured?

apelisse replied:
I haven't looked extensively, but it's possible that we actually run the FieldManager on the scale object itself before :-(.

nodo replied:
It should be quite safe, `options` is created [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-c4afbb7d6a8bdb1d6475f18abc92caa6c413f879aeeb189187c569f3f6c146f4R584). Do you prefer me to add extra checks?

nodo replied:
Yes, if I am not mistaken, it's used on the scale object itself too (e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L553)

apelisse replied:
Yeah, for nothing, but yes :D

apelisse replied:
Yeah, that's a good point, thanks for mentioning that.

@nodo, can we start by adding a test that reproduces that problem. That is:
- create a deployment,
- Run "Apply" with a "Scale" object that has a different number for replicas, without "force", and make sure we have a conflict.

My understanding is that 2. is mostly what we already do, we need to implement 1. to fix the test described above?

nodo replied:
Right, I have added the test here: https://github.com/kubernetes/kubernetes/pull/95921/commits/2c1408803f8d33cfbf2d331ac594380e788f6795
and it works as expected.

If I understand correctly. The field manager of the main resource is invoked also when running "apply on "scale" so it triggers a conflict. Am I missing something though?

apelisse replied:
Ah! There's something tricky happening here. To verify that, I think you could write a test where apply will actually ALWAYS conflict on scale. When you apply an object that doesn't have a managedFields set (and that's the case of EVERY scale objects here), we create one that we call "before-first-apply" with existing fields. Here, you're getting a conflict with that manager, not the actual manager that changed the value. Every time you apply, we'll re-create this "before-first-apply" manager (even though you're supposed to own the field), and you'll get a conflict.

I don't know yet how to solve that.

nodo replied:
Thanks a lot for spotting, that's quite a tricky edge case :)

I think a possible solution would be to not run `Apply` on scale objects. E.g. before [this line](https://github.com/nodo/kubernetes/blob/ssa-scale-ownership-deployment/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L441)

Add something like this:

```go
if scale, ok := patchObject.(*autoscaling.Scale); ok {
  return scale, nil
}
```
Applying the field manager here doesn't have any effect anyway.

I tried to implement it but I got stuck because I couldn't find a way to convert an `*unstructured.Unstructured` to `*autoscaling.Scale` and I wanted to check with you first.

What do you think about this solution? If it's a good way forward, how to do the conversion?

nodo replied:
I have done some more tests and I was trying to convert the `Unstructured` object using the following code:

```go
scaleGVK := schema.GroupVersionKind{Group: "autoscaling", Version: "v1", Kind: "Scale"}
if p.kind == scaleGVK {
        var scale autoscaling.Scale
        err := runtime.DefaultUnstructuredConverter.FromUnstructured(patchObj.UnstructuredContent(), &scale)
        return &scale, err
}
```

This doesn't work for a reason I don't fully understand. The resulting `scale` object doesn't have any metadata, and I am getting this error:
```
error the name of the object (deployment based on URL) was undeterminable: name must be provided
```

Digging a bit deeper, I have noticed that [the Scale object](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/types.go#L28-L41) does not specify any JSON annotations, why is that? Other types such as [Deployment](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1/types.go#L254-L267) have them. Could this be a problem?

apelisse replied:
> Digging a bit deeper, I have noticed that the Scale object does not specify any JSON annotations, do you know the reason for that? Other types such as Deployment have them. Could this be a problem?

You're looking at the "internal type" rather than the "external type". Internal types are not converted to json and hence don't have the tags. You should be looking at the external type (note the difference in the URL, it's probably too subtle ...): https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/autoscaling/v1/types.go#L114-L127

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -279,10 +280,11 @@ func (r *ScaleREST) Get(ctx context.Context, name string, options *metav1.GetOpt
 		ctx,
 		name,
-		&scaleUpdatedObjectInfo{name, objInfo},
+		&scaleUpdatedObjectInfo{name, objInfo, parentFieldManager, options.FieldManager},
```
Are we guaranteed that `options` is always non-nil here?

nodo replied:
It should be quite safe, options is created here, just before the Update call. Do you prefer me to add extra checks?

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -401,8 +405,15 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 	}
 
-	// move replicas/resourceVersion fields to object and return
+	live := deployment.DeepCopy()
```
I'm assuming that we need to copy in case the `Update` changes the object in place. I'm pretty sure it doesn't (do you remember @jpbetz?). Either way, there is no reason to copy if we don't call `Update`, at least move the copy in the if statement. But I don't think it's needed.

apelisse replied:
I might be wrong, have you experimented with it?

nodo replied:
My assumption was that we need to pass to `Update` two objects: 1. the object "before" changing it and 2. the object "after" changing it so that the managed fields are updated accordingly. If we don't copy it we effectively pass the same object. I am probably missing something though. 

If we need to keep the copy, and we decide we want to do it only when necessary, the resulting code is a bit convoluted. I would propose something like:
```go
	var live *apps.Deployment
	if i.parentFieldManager != nil {
		live = deployment.DeepCopy()
	}

	// move replicas/resourceVersion fields to object
	deployment.Spec.Replicas = scale.Spec.Replicas
	deployment.ResourceVersion = scale.ResourceVersion

	if i.parentFieldManager != nil {
		// update managed fields using the fieldmanager of the main resource
		return i.parentFieldManager.Update(live, deployment, i.manager)
	}
	return deployment, nil
```
WDYT?

apelisse commented:
LGTM once we understand if we need to copy the object.

apelisse commented:
/lgtm
/approve

apelisse commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2811,3 +2811,192 @@ spec:
+	assertReplicasOwnership(t, (*deployment).GetManagedFields(), "replace_test")
+}
+
+func assertReplicasOwnership(t *testing.T, managedFields []metav1.ManagedFieldsEntry, fieldManager string) {
```
I'm really not a big fan of this here. Other suggestion:
1. Check that there is an entry for the manager used in the scale operation,
2. Check for conflicts on changes to that field

nodo replied:
I am not sure I fully understood.

> Check that there is an entry for the manager used in the scale operation,

This is done [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR2992-R2994) unless I missed something. `fieldManager` is the manager used in the scale operation, see for instance [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR2896).

> Check for conflicts on changes to that field

[This](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR2898-R2910) was my attempt to check for conflict, since the checks happens right after applying I thought in made sense to check it outside an helper.

nodo replied:
I couldn't find an easy we have a way to check the that `replicas` field is owned by a field manager (and not others) without decoding. I am open to suggestions though.

I see your point of not checking internal structures, but we don't have unit tests for this feature and I would feel more confident having a more strict integration tests. As we discussed, we could definitely refactor this bit into a clearer framework.

WDYT?

apelisse commented:
/lgtm
/assign @liggitt

nodo commented:
Thanks a lot @liggitt for review! 🙏 

>Does that mean that even if this PR enabled tracking ownership via /scale, server-side-applying to /scale would not notify of field manager conflicts on the replicas field? Shouldn't it?

Field manager will be notified on conflicts (see the integration test [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR2898-R2910)). It just runs the logic of the deployment field manager.


>Rather than trying to run two fieldmanager passes on Scale, could we: [...]

The problem with this solution is that we would need to merge back the `scale` managed field into the deployment one. This is not trivial because we need to update the existing entries that owns `replicas` and remove it. In other words, if "replicas" is owned by manager A, we run `/scale` using manager B. We want `replicas` to be removed from A and assigned to B.

That's why at the end this solution seemed more straightforward.

nodo commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -401,8 +405,15 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
+
+	if i.parentFieldManager != nil {
+		// update managed fields using the fieldmanager of the main resource
+		return i.parentFieldManager.Update(live, deployment, i.manager)
```
That would be my assumption, but I am not 100% sure. @apelisse do you know?

apelisse replied:
This is in apps/deployment, so I'm assuming the question is unrelated to the change?

With regard to the question, yes, we would do the right thing. "managedFields" are tracked per version, and we conflict when the value changes, which we can detect by converting the object to the version of the managedFields.

liggitt replied:
yeah, sorry... the question was about what would happen if we extrapolated this approach to the scale subresource for a CRD

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go:
```diff
@@ -171,6 +172,8 @@ func UpdateResource(r rest.Updater, scope *RequestScope, admit admission.Interfa
 		trace.Step("About to store object in database")
 		wasCreated := false
 		requestFunc := func() (runtime.Object, error) {
+			ctx = fieldmanager.ToContext(ctx, scope.ParentFieldManager)
```
I'm not personal the most knowledgeable about wiring in the apiserver in general, but I had the same reaction when I saw that change. I've tried to find a better way and couldn't find one, that of course doesn't mean there isn't one.

nodo committed:
Add test when running "Apply" on "/scale"

1. It produces a conflict when .spec.replicas is different from the main
resource
2. It updates the deployment when forcing it

nodo committed:
Add more tests around applying directly scale objects

liggitt commented:
> > Does that mean that even if this PR enabled tracking ownership via /scale, server-side-applying to /scale would not notify of field manager conflicts on the replicas field? Shouldn't it?
> 
> Field manager will be notified on conflicts (see the integration test [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR2898-R2910)). It just runs the logic of the deployment field manager.

Does that conflict error come from here:
```
		// update managed fields using the fieldmanager of the main resource
		return i.parentFieldManager.Update(live, deployment, i.manager)
```

Do `update deployments/scale` and `apply-patch deployments/scale` behave as expected (apply-patch can return field manager conflicts, update always stomps)?

nodo commented:
@liggitt you raised a very good points in your latest review. The conflict are not raised properly when applying a `autoscaling/v1 Scale` object as they are raised [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L441). I still haven't found a way forward, but we had a few discussion with @apelisse and @kwiesmueller .

nodo committed:
WIP

nodo commented:
Closing in favour of https://github.com/kubernetes/kubernetes/pull/98377
